notes/compilation-and-linkage.md

```diff
@@ -0,0 +1,684 @@
+
+<div style="text-align: center">
+
+`#if defined DIRECTIVE | #if !defined DIRECTIVE`
+
+or
+
+`#if DIRECTIVE == 0 | #if DIRECTIVE != 0`
+
+or
+
+`#ifdef DIRECTIVE | #ifndef DIRECTIVE`
+
+</div>
+
+The last one is the most common one to do since it is the most condensed. It works just like any conditional, but only in the preprocessor is the operator `defined` introduced. It is done so to accommodate this kind of test. 
+
+A word on these options:
+1. The last option is the condensed form of the first, so when the preprocessor reads it, it converts to the first option.
+2. The second option is a consequence of the lack of directive types. Non-existent or undefined directives point to a null, or zero space.
+3. This operation only works on macros, or preprocessor directives using `#defined`.
+
+### Header Files
+
+There are a few issues with pure C/C++ files:
+
+1. Including other C/C++ files can be bad, especially if they are large, as the preprocessor will copy and paste those files
+2. If we are designing a system, how do we ensure it is complete? In other words, how do we make sure all classes/functions are implemented?
+3. Trying to avoid the first problem can cause us to `#include` the same thing in multiple files. Ideally, there should be as little places to repeat directives.
+
+We can solve this with header files (.h/.hpp). These are also C/C++ files, and technically both .h/.hpp and .c/.cpp files can do the same thing, with the exception that explicit calls to the compiler can only occur on .c/.cpp files. 
+
+So what's the point? Header files are special C/C++ files where we state all preprocessor, class, function, and type definitions. The class and function definitions are then implemented by the corresponding .c/.cpp file, usually of the same name. Its a little bit like a Java interface, if that helps at all, except its not strictly for a class.
+
+Let's look at an example implementation. Here we have `point.hpp`, `point.cpp`, and `pointMain.cpp`
+
+```C++
+#define SHIFT_X = 4;
+#define SHIFT_Y = 5;
+
+#define <iostream>
+#define <string>
+
+struct point {
+    std::string name;
+    int x;
+    int y;
+};
+
+point DefaultTranslate();
+
+point Translate(std::string name, int xShift, int yShift);
+
+void printString(point p);
+```
+<div style="text-align: center">
+
+Figure 6a - `point.hpp`
+
+</div>
+
+```C++
+#include "point.hpp"
+
+point Translate(std::string name, int xShift, int yShift) {
+    if(xShift == 0 && yShift == 0) {
+        return DefaultTranslate(name);
+    } else {
+        return {name, x + xShift, y + yShift};
+    }
+}
+
+point DefaultTranslate(std::string name) {
+    return {name, x + SHIFT_X, y + SHIFT_Y};
+}
+
+void printString(point p) {
+    std::cout << Coordinate Points for << p.name << ": (" << x << ", " << y << ")" << std::endl;;
```

Comment:
Need the double quotes around the string, make sure stuff compiles if people decide to copy/paste